client/app/components/EditParameterSettingsDialog.jsx

```diff
@@ -0,0 +1,204 @@
+import { includes, words, capitalize, clone, isNull } from 'lodash';
+import React, { useState, useEffect } from 'react';
+import PropTypes from 'prop-types';
+import Modal from 'antd/lib/modal';
+import Form from 'antd/lib/form';
+import Checkbox from 'antd/lib/checkbox';
+import Select from 'antd/lib/select';
+import Input from 'antd/lib/input';
+import Divider from 'antd/lib/divider';
+import { wrap as wrapDialog, DialogPropType } from '@/components/DialogWrapper';
+import { Query } from '@/services/query';
+
+const { Option } = Select;
+const formItemProps = { labelCol: { span: 6 }, wrapperCol: { span: 16 } };
+
+function useQuerySearch() {
+  const [searchTerm, setSearchTerm] = useState(); // search term
+  const [searchResults, setSearchResults] = useState([]); // search results
+
+  useEffect(() => {
+    if (!searchTerm || searchTerm.length < 3) {
+      return;
+    }
+
+    Query.query({ q: searchTerm }, ({ results }) => {
+      setSearchResults(results);
+    });
+  }, [searchTerm]); // search when term changes
+
+  return [searchResults, setSearchResults, setSearchTerm];
+}
+
+function useParamSave(onConfirm, onClose) {
+  const [saving, setSaving] = useState(false);
+
+  const save = (param) => {
+    setSaving(true);
+    onConfirm(param)
+      .then(() => {
+        onClose(param);
+      }).finally(() => {
+        setSaving(false);
+      });
+  };
+
+  return [saving, save];
+}
+
+function getDefaultTitle(text) {
+  return capitalize(words(text).join(' ')); // humanize
+}
+
+function useNewName(existingParams) {
+  const [name, setName] = useState('');
+  const [feedback, setFeedback] = useState({ isError: false, text: null });
+
+  useEffect(() => {
+    const alreadyExists = includes(existingParams, name);
+    if (alreadyExists) {
+      setFeedback({ isError: true, helpText: 'Parameter with this name already exists' });
+    } else if (!name) {
+      setFeedback({ isError: false, helpText: 'Choose a keyword for this parameter' });
+    } else {
+      setFeedback({ isError: false, helpText: `This is what will be added to your query editor {{ ${name} }}` });
+    }
+  }, [name]);
+
+  return [feedback, name, setName];
+}
+
+function EditParameterSettingsDialog(props) {
+  const [param, setParam] = useState(clone(props.parameter));
+  const [querySearchResults, setQuerySearchResults, setQuerySearchTerm] = useQuerySearch();
+  const [newNameFeedback, newName, setNewName] = useNewName(props.existingParams);
+  const [saving, save] = useParamSave(props.onConfirm, props.dialog.close);
+  const isNew = !props.parameter.name;
+
+  // get query name by id
+  useEffect(() => {
+    if (param.queryId) {
+      Query.get({ id: props.parameter.queryId }, (query) => {
+        setQuerySearchResults([query]);
+      });
+    }
+  }, []);
+
+  // keep name in sync
+  useEffect(() => {
+    setParam({ ...param, name: newName });
+  }, [newName]);
```

Comment:
In the CodeSandBox example above, I also removed the need for a separate state for the name, so not sure this is needed.